### PR TITLE
Implemented Missing Bits of JSP 3.1 Specification

### PR DESCRIPTION
**Pull request by ropalka**
https://issues.redhat.com/browse/UNDERTOW-2106
https://issues.redhat.com/browse/UNDERTOW-2107
https://issues.redhat.com/browse/UNDERTOW-2108

### CODE REVIEW

File: src/main/java/org/apache/jasper/el/ELResolverImpl.java [hunk 1]
```diff
@@ -66,9 +65,7 @@ public Object getValue(ELContext context, Object base, Object property) {
 
     @Override
     public Class<?> getType(ELContext context, Object base, Object property) {
-        if (context == null) {
-            throw MESSAGES.elResolverNullContext();
-        }
+        Objects.requireNonNull(context);
```
Comment (fl4via) [inline src/main/java/org/apache/jasper/el/ELResolverImpl.java]: Whenever replacing a Messages.XXX by Objects.requireNonNull, we need to add the logging message, and, I'm not sure if possilble (probably not) we should pass the error id and keep it unchanged

File: src/main/java/org/apache/jasper/runtime/JspApplicationContextImpl.java [hunk 2]
```diff
@@ -46,95 +46,94 @@ public class JspApplicationContextImpl implements JspApplicationContext {
...
-
-	public ELContextImpl createELContext(JspContext context) {
-		if (context == null) {
-			throw MESSAGES.nullJspContext();
-		}
-
-		// create ELContext for JspContext
+    public void addELContextListener(ELContextListener listener) {
+        if (listener == null) {
+            throw MESSAGES.nullElContextListener();
```
Comment (fl4via) [inline src/main/java/org/apache/jasper/runtime/JspApplicationContextImpl.java:67]: Since you started replacing this by Objects.requiresNonNull maybe you should keep doing it here. And, ideally, these changes should be in a separate commit, having changes that update code and/or fix style at the same commit that adds and or changes functionality makes reviewing the PR harder

**Review by fl4via — CHANGES_REQUESTED** [review]
Overall it looks very good. Only minor comment about Object.requiresNonNull changes.

File: src/main/java/org/apache/jasper/el/ELResolverImpl.java (hunk 1 above)
Comment (ropalka) [reply src/main/java/org/apache/jasper/el/ELResolverImpl.java]: I reverted this change. This was the result of the sync with the Apache Tomcat's code base.

File: src/main/java/org/apache/jasper/runtime/JspApplicationContextImpl.java (hunk 2 above)
Comment (ropalka) [reply src/main/java/org/apache/jasper/runtime/JspApplicationContextImpl.java:67]: I didn't do it intentionally @fl4via . All these changes were the result of the sync process with part of Apache Tomcat's code base.

**Comment by ropalka** [comment]
I fixed all your comments @fl4via .
